src/Corax/Pipeline/Parsing/ScalarTokenizers.cs

```diff
@@ -0,0 +1,260 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Corax.Pipeline.Parsing
+{
+    internal static class ScalarTokenizers
+    {
+        public static int TokenizeWhitespaceAsciiScalar(ReadOnlySpan<byte> buffer, ref Span<Token> tokens)
+        {
+            ref byte bufferStart = ref MemoryMarshal.GetReference(buffer);
+
+            int size = 0;
+            int tokenIdx = 0;
+            
+            int idx = 0;
+            for (; idx < buffer.Length; idx++)
+            {
+                byte b = Unsafe.Add(ref bufferStart, idx);
+
+                if (b > 0x20 || ((ScalarParsers.SingleByteTable >> b) & 1) == 0)
+                {
+                    size++;
+                    continue;
+                }
+
+                if (size == 0)
+                    continue;
+
+                ref var token = ref Unsafe.Add(ref MemoryMarshal.GetReference(tokens), tokenIdx);
+                token.Offset = idx - size;
+                token.Length = (uint)size;
+                token.Type = TokenType.Ascii | TokenType.Word;
+                tokenIdx++;
```

Comment:
How do you protect against out of range accesses here? We don't check `tokenIdx < tokens.Length`

Reply:
`Token` is mapped 1-to-1 to characters. there is no way we can have a range access violation.

Reply:
Nonetheless I will add debug time checks on that, to ensure caller does not violate that invariant.